app/upgrades/v13/upgrades.go

```diff
@@ -1,15 +1,84 @@
 package v13
 
 import (
+	"fmt"
+	"os"
+	"strings"
+
+	wasmkeeper "github.com/CosmWasm/wasmd/x/wasm/keeper"
+	wasmtypes "github.com/CosmWasm/wasmd/x/wasm/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 	"github.com/cosmos/cosmos-sdk/types/module"
+	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
 	upgradetypes "github.com/cosmos/cosmos-sdk/x/upgrade/types"
-
+	transfertypes "github.com/cosmos/ibc-go/v3/modules/apps/transfer/types"
 	"github.com/osmosis-labs/osmosis/v12/app/keepers"
 	"github.com/osmosis-labs/osmosis/v12/app/upgrades"
+	ibcratelimittypes "github.com/osmosis-labs/osmosis/v12/x/ibc-rate-limit/types"
 	lockuptypes "github.com/osmosis-labs/osmosis/v12/x/lockup/types"
 )
 
+func setupRateLimiting(ctx sdk.Context, keepers *keepers.AppKeepers) error {
+	govModule := keepers.AccountKeeper.GetModuleAddress(govtypes.ModuleName)
+	code, err := os.ReadFile("rate_limiter.wasm")
+	if err != nil {
+		return err
+	}
+	contractKeeper := wasmkeeper.NewGovPermissionKeeper(keepers.WasmKeeper)
+	instantiateConfig := wasmtypes.AccessConfig{Permission: wasmtypes.AccessTypeOnlyAddress, Address: govModule.String()}
+	codeID, err := contractKeeper.Create(ctx, govModule, code, &instantiateConfig)
+	if err != nil {
+		return err
+	}
+
+	transferModule := keepers.AccountKeeper.GetModuleAddress(transfertypes.ModuleName)
+
+	denoms := []string{
+		"uosmo",
+		"ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2", // atom
+		"ibc/D189335C6E4A68B513C10AB227BF1C1D38C746766278BA3EEB4FB14124F1D858", // usdc
+		"ibc/EA1D43981D5C9A1C4AAEA9C23BB1D4FA126BA9BC7020A25E0AE4AA841EA25DC5", // weth
+		"ibc/EA1D43981D5C9A1C4AAEA9C23BB1D4FA126BA9BC7020A25E0AE4AA841EA25DC5", // wbtc
+	}
+
+	quotas := ""
+	for _, denom := range denoms {
+		quotas += fmt.Sprintf(`
+{"channel_id": "any", "denom": "%s", "quotas": [{"name":"weekly", "duration": 604800, "send_recv":[50, 50]}] },`, denom)
+	}
+	if quotas != "" {
+		// Remove the trailing comma because JSON can't handle it
+		quotas = strings.TrimRight(quotas, ",")
+	}
+
+	initMsgBz := []byte(fmt.Sprintf(`{
+           "gov_module":  "%s",
+           "ibc_module":"%s",
+           "paths": [%s]
+        }`,
+		govModule, transferModule, quotas))
```

Comment:
What is the transferModule allowing here? Did we decide to remove this from the smart contract or are we leaving it in for a particular reason?

Reply:
I was gonna remove it and then found why I had added that in the first place. The transfer module is allowed to add or remove quotas. The idea is that in the future we may want to automatically add or remove channels as they get created/deleted